### ISSUE TITLE
![image](https://user-images.githubusercontent.com/8941351/54089769-d7f1fc80-436c-11e9-837d-45cd2ab83508.png)

### ISSUE THREAD

**Issue by mvburgh** [zeromq/netmq#788]
![image](https://user-images.githubusercontent.com/8941351/54089769-d7f1fc80-436c-11e9-837d-45cd2ab83508.png)

my code is already minimalistic; only a RequestSocket and ResponseSocket.
I'm creating a RequestSocket per few seconds based on the requests from a Web Controller

_Originally posted by @mvburgh in https://github.com/zeromq/netmq/issues/737#issuecomment-471332420_

**Comment by KamranShahid** [comment]
Any one working on it?

**Comment by mvburgh** [comment]
Not me personally. I had a quick glance at the code but could not see any possible cause in the meanwhile.

**Comment by Svisstack** [comment]
![image](https://user-images.githubusercontent.com/864295/63580362-f3586900-c594-11e9-8533-f0ba78da9299.png)

@mvburgh Did you made any progress since then on that?

**Comment by somdoron** [comment]
I will take a look next week, on vacation this week.

Code that reproduce this will help

**Comment by somdoron** [comment]
Also @svisstack, can you check with socket.Options.Linger set to zero?
I suspect it might be the issue.

**Comment by Svisstack** [comment]
![image](https://user-images.githubusercontent.com/864295/63584247-d162e480-c59c-11e9-8480-9f4bd1532964.png)

Code which causing this is very simple just the PublisherSocker who had connected ~10 clients:

![image](https://user-images.githubusercontent.com/864295/63584373-1c7cf780-c59d-11e9-9af7-f2f10a5f3e39.png)

I'm not sure it's the same bug as I'm not sure it's linked to the ResponseSocket internally. 

From the dumps, we can see that there are probably too many Pub+PubSession objects and along with that Pipe, YPipe, YQueue, but all the memory is allocated on the YQueue+Chunk<Msg>

**Comment by Svisstack** [comment]
@somdoron I confirm that Linger is equal to the {00:00:00} at the end of the Start() function in the snippet provided above @(Start(): return port;)

publisher.Options
{NetMQ.SocketOptions}
    Affinity: 0
    Backlog: 100
    DelayAttachOnConnect: false
    DisableTimeWait: false
    Endian: Big
    IPv4Only: true
    Identity: null
    LastEndpoint: "tcp://0.0.0.0:61584"
    LastPeerRoutingId: null
    Linger: {00:00:00}
    MaxMsgSize: -1
    MulticastHops: 1
    MulticastRate: 100
    MulticastRecoveryInterval: {00:00:10}
    PgmMaxTransportServiceDataUnitLength: 'publisher.Options.PgmMaxTransportServiceDataUnitLength' threw an exception of type 'NetMQ.InvalidException'
    ReceiveBuffer: 0
    ReceiveHighWatermark: 1000
    ReceiveLowWatermark: 0
    ReceiveMore: false
    ReconnectInterval: {00:00:00.1000000}
    ReconnectIntervalMax: {00:00:00}
    SendBuffer: 0
    SendHighWatermark: 0
    SendLowWatermark: 0
    TcpKeepalive: false
    TcpKeepaliveIdle: {-00:00:00.0010000}
    TcpKeepaliveInterval: {-00:00:00.0010000}

**Comment by somdoron** [comment]
Do the subscribers come and go frequently?
It seems like linger set to zero or few seconds will solve

**Comment by somdoron** [comment]
Thanks, does the subscribers come and go? 
Can you check who is referencing the PubSession?

**Comment by Svisstack** [comment]
@somdoron Take a look at the incoming reference chart.

![image](https://user-images.githubusercontent.com/864295/63586037-bdb97d00-c5a0-11e9-9a32-8b7902745b75.png)

In my use-case, the subscribers should not come and go frequently, but there could be a bug on my side causing the come and go and I analyzing that at the moment.

**Comment by Svisstack** [comment]
I'm using the 4.0.0.239-pre version.

**Comment by somdoron** [comment]
Can you send me the report? Which application are you using? 

somdoron AT gmail DOT com

I'm not in front of a computer this week, but I will take a look beginning of next week.

**Comment by Svisstack** [comment]
@somdoron No problem, I actually found interesting fact - the leak is visible only on nodes on which there is no communication activity between Publisher and Subscriber (silence), it's ok from the application perspective.

**Comment by somdoron** [comment]
Can you extend this list:

https://user-images.githubusercontent.com/864295/63584247-d162e480-c59c-11e9-8480-9f4bd1532964.png

I want to see the root object causing the memory leak

**Comment by somdoron** [comment]
Also, can you show the incoming reference to the pipe class?

**Comment by Svisstack** [comment]
![image](https://user-images.githubusercontent.com/864295/63587677-e6437600-c5a4-11e9-89c4-43efa0b168e9.png)

It looks like the Pipe is also referenced to the Pub+Sub, however, I don't know it's the same instance.

**Comment by Svisstack** [comment]
![image](https://user-images.githubusercontent.com/864295/63587942-8bf6e500-c5a5-11e9-86a2-1542c1ad0f72.png)

@somdoron Visual Studio detected multiple reference cycles in the paths to root.

**Comment by somdoron** [comment]
Funny, I just figured it out myself.

At least in this case it is not a bug.

Once one message will be sent everything will be freed.

From the memory picture I saw pending command holding the reference and causing the issue.

To avoid the issue you can call once in a while the socket.Poll with zero timespan. This will also process pending commands.

Anyway, I think you have a case where subscribers come and go frequently.

**Comment by Svisstack** [comment]
Thanks. @somdoron, I appreciate the effort and in-depth knowledge about the source of this project.

Have a nice time on the vacations.

Yes, I could have the come and go issue looking at the netstat.

**Comment by mvburgh** [comment]
> Do the subscribers come and go frequently?
> It seems like linger set to zero or few seconds will solve

In my case they come and go every few seconds as they are web api requests.

**Comment by somdoron** [comment]
@mvburgh, i will try to reproduce next week.
Only request response sockets? Are you using a proxy? Do you happen to have memory profiler report?

**Comment by mvburgh** [comment]
No proxy here; it runs between a windows service and website for me.
I dont have a profile report at hand.

**Comment by mvburgh** [comment]
@somdoron did you get any further? Is there anything you need from me?

**Comment by ReneOlsthoorn** [comment]
Same memoryleak in our product. We use netmq to distribute small data to a lot of consumers. Connecting, disconnecting and reconnecting happens a lot. Our product ends up with 3Gb of memory and crashes. At start, it uses 30MB of memory. We use the publish and subscribe classes in netmq. I did investigate that the problem was not somewhere else. Just connecting and disconnecting in a rappid pace is enough to get a memoryleak.

**Comment by KamranShahid** [comment]
I have majordomo pattern implemented with broker in one windows service (.net core 2.1) and worker app resides on another windows service (.net core 2.1).
In worker windows service there are different 16/17 type of workers . each type of worker can have multiple instances. What i were seeing is when i am assigning 10 number of worker against each type my broker application memory increases time to time.

It probably is due to default heartbeat time. Now I am trying setting default heartbeat time at worker side as 10 seconds while on broker 15 seconds.

**Comment by somdoron** [comment]
@ReneOlsthoorn during the time the memory increase to 3Gb are you still sending messages? can it be that it happens only during silence times?

**Comment by ReneOlsthoorn** [comment]
@somdoron Yes, the server keeps running and messages are send. During silence times no memory is increased. The memory increase is gradually, every 4 hours one Gb. It depends how much consumers are connecting and disconnecting. I've made a test-program which connects and disconnects. The memoryleak is visible there as well. I've cloned the git sources, so maybe I can see where the problem is.

**Comment by somdoron** [comment]
Can you share the test program?

**Comment by ReneOlsthoorn** [comment]
Unfortunately, the test program is full of proprietary code. If the priority is high enough for our organisation, I will debug netmq and maybe find out if it's a bug or not. I will keep you all posted in this issue.

**Comment by somdoron** [comment]
Can you share a memory profiler snapshot? that will help alot

**Comment by ReneOlsthoorn** [comment]
Doron and others, the memory-leak I was investigating was in our own product. My apologies for posting when it was not clear where the problem came from. I've deleted my comments, so new users don't get a wrong impression about NetMQ.
Keep up the good work!

**Comment by mvburgh** [comment]
@somdoron i kind of lost track on this item for a while. Is there anything you need from me? Or can i work around something with the messages above?

**Comment by mvburgh** [comment]
@somdoron I have spent some more time with this last week, but both setting the linger to 0 and the socket.Poll() every now and then give no better result. The increase stays in YQueue+Chunk and does not get freed over time.

**Comment by manu-st** [comment]
We are also experiencing something similar in our app. We do not see a leak when we have one server and one client communicating via Request/Response sockets. However, if another client tries to connect to the server while it is already serving another client, the server will leak memory. The way we have it work is that the server can only serve one client, so when a new client connects, it sends a message to tell the client that it cannot communicate and that's pretty much it. Once the client receives that message it disconnects.
